File: ~Programming_cheat_sheets/_compiled_languages/Java/_data-structures-and-algorithms/__sorting-algorithms/HeapSort.java
package com.example;

import java.util.Arrays;

class HeapSort {
    private int[] array;

    public HeapSort(int[] array) {
        this.array = array;
    }

    public void sort() {
        if (array.length < 2) return;

        // Create heap
        for (int i = 0; i < array.length; i++) {
            bubbleUp(i);
        }

        // Swap highest values to end or array
        for (int i = 0; i < array.length; i++) {
            int currLastIndex = array.length - 1 - i;
            swap(0, currLastIndex);
            bubbleDown(currLastIndex);
        }
    }

    //
    // Helpers
    //

    private int PI(int i){ return (i-1)/2;}
    private int LI(int i){ return 2 * i + 1;}
    private int RI(int i){ return 2 * i + 2;}
    private boolean hasP(int i){ return i > 0;}
    private boolean hasL(int i, int size){ return LI(i) < size;}
    private boolean hasR(int i, int size){ return RI(i) < size;}
    private int PV(int i){ return array[PI(i)]; }
    private int LV(int i){ return array[LI(i)]; }
    private int RV(int i){ return array[RI(i)]; }

    private void bubbleUp(int i) {
        while (i >= 0 && hasP(i) && PV(i) < array[i]) {
            swap(PI(i), i);
            i = PI(i);
        }
    }

    // currLastIndex will NOT be touched/swapped
    private void bubbleDown(int currLastIndex) {
        int i = 0;

        while (hasL(i, currLastIndex)) {
            int gci = LI(i);
            if (hasR(i, currLastIndex) && RV(i) > LV(i)) {
                gci = RI(i);
            }

            if (array[i] < array[gci]) {
                swap(i, gci);
                i = gci;
            } else {
                break;
            }
        }
    }

    private void swap(int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

public class HeapSortExample {
    public static void main(String[] args) {
        int[] arr = {5, 4, 6, 3, 7, 2, 8, 1, 0, 9};
        System.out.println("Arrays.toString(arr) = " + Arrays.toString(arr));
        HeapSort heapSort = new HeapSort(arr);
        heapSort.sort();
        System.out.println("Arrays.toString(arr) = " + Arrays.toString(arr));
    }
}

/* OUTPUT

Arrays.toString(arr) = [5, 4, 6, 3, 7, 2, 8, 1, 0, 9]
Arrays.toString(arr) = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

*/

File: ~Programming_cheat_sheets/_compiled_languages/C/_data-structures-2020/sorting-algorithms/heap-sort/heap-sort.c
//
// Heap sort
//

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

// BE CAREFUL, if these are wrong, they can do a lot of damage
#define length(x) (sizeof(x)/sizeof(x[0]))
#define getPI(i) ((i - 1) / 2)
#define getLI(i) ((2 * i) + 1)
#define getRI(i) ((2 * i) + 2)
#define getPV(a, i) (a[getPI(i)])
#define getLV(a, i) (a[getLI(i)])
#define getRV(a, i) (a[getRI(i)])
#define hasP(a, i) (i != 0 && a[getPI(i)] >= 0)
#define hasL(i, size) (getLI(i) < size)
#define hasR(i, size) (getRI(i) < size)

//
// Functions
//

void printArray(int a[], int size){
  printf("Array@%p [", a);
  for(int i=0;i<size;i++){
    if (i < size-1) printf("%d, ", a[i]);
    else printf("%d", a[i]);
  }
  printf("]\n");
}

void swap(int a[], int i1, int i2){
  int temp = a[i1];
  a[i1] = a[i2];
  a[i2] = temp;
}

void bubbleDownUntil(int a[], int size){
  int idx = 0;
  while (hasL(idx, size)) {
    int gci = getLI(idx);
    if (hasR(idx, size) && getRV(a, idx) > getLV(a, idx)) {
      gci = getRI(idx);
    }

    if (a[gci] > a[idx]){
      swap(a, gci, idx);
      idx = gci;
    }else{
      break;
    }
  }
}

void bubbleUp(int a[], int idx){
  while (hasP(a, idx) && getPV(a, idx) < a[idx]) {
    swap(a, getPI(idx), idx);
    idx = getPI(idx);
  }
}

// We build max heap, then swap index 0 with current-last
void heapSort(int a[], int size){
  for(int i=1;i<size;i++){
    bubbleUp(a, i);
  }

  for(int i=(size-1);i>=1;i--){
    swap(a, 0, i);
    bubbleDownUntil(a, i); // las-index (i) should NOT be touched!
  }
}

//
// Main function
//

int main(){
  int arr[] = {
    5,4,6,3,7,2,8,1,9,0
  };

  heapSort(arr, length(arr));
  printArray(arr, length(arr));
  return 0;
}

/* OUTPUT

Array@0x7ffeee8a7aa0 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

*/
